Christoph country

Christoph98 opened the pull request:
Temporarily fixed the get-organizations API.

thyo9470 commented on code/routers/country_page/country.js:
```diff
@@ -101,10 +101,10 @@ router.get('/get-country-info', (req, res) => {
       countryName: String
 
-    Returns the list of organizations associated with given country
+    Returns the list of organizationIDs associated with given country
```
Please add a TODO here to note that this is not finished and that it will be changed back to organizations

Christoph98 replied:
Now updated it, so I didn't add a todo.

thyo9470 commented on code/routers/country_page/country.js:
```diff
@@ -99,28 +99,192 @@ router.get('/get-country-info', (req, res) => {
+    }else{
+      var updatedOrgIDs = []
+      Country.find(countryQuery, function(err, country) {
+        if(err) {
```
This should check if the length of country is > 0 otherwise line 176 can cause an error

thyo9470 commented on code/routers/country_page/country.js:
```diff
@@ -99,28 +99,192 @@ router.get('/get-country-info', (req, res) => {
+    }else{
+      var updatedOrgIDs = []
+      Country.find(countryQuery, function(err, country) {
+        if(err) {
```
This should also check if country length > 0 or else line 250 can cause an error

thyo9470 commented on code/routers/country_page/country.js:
```diff
@@ -99,28 +99,192 @@ router.get('/get-country-info', (req, res) => {
+              return toKeep
+            })
+          }
+          // Normally doing this:
+
+          //country.organizationIDs = updatedOrgIDs;
+          //country.save();
+
+          // would be better to update country, but country is not a mongoose.Document
+          // this is likely due to how the data was imported on the mongoDB Atlas database
```
1. Is this comment necessary?
2. Why is it not like this? If this is the optimal solution that we should try and make this work so that the code is simpler and thus less likely to cause bugs

Christoph98 replied:
I see what you mean, it would be interesting to see this method work (and it is theoretically better, idk why, but mongoose documentation says so), but since this isn't a user-facing API, I think it isn't worth the effort to make it work. The problem with the `save()` function is that it exists for `mongoose.Documents`, but when doing a mongoose `Country.find(query, function(err, country){})`, country is not a document. I believe this has to do with how I imported the data to the database. Unfortunately I don't have the time to go back and try using mongoose (as opposed to directly calling `mongoimport`) to populate the database. 

I can delete the comment if you think I should, but I think it would be good to keep for the future team.

thyo9470 replied:
The find method should return some json which I believe you can easily convert into a mongoose document doing something like this:

var organization = new Organization([JSON returned from find])

This assumes that it only returns one object, if it returns more you will have to loop over the list and convert each one.

Also, comments like this should be put in the wiki, it is bad practice to put long comments like this in the code. Generally the code should only have brief comments for things that aren't obvious. This should be put int the wiki and have a link to it in the code with a brief thing saying look at this wiki page for information. Before doing this try the thing above.

Christoph98 commented:
- I've added my changes to the data folder to this branch
- I've implemented `country.length` checks as recommended
- I've added a 'noCountryOrganizations' error to the `remove-organizations` API

thyo9470 commented on code/data/create_electricity_json.py:
```diff
@@ -31,6 +32,24 @@ def list_to_string(l):
 
+#additional countries to add
+stats = [NaN for i in range(1990, 2017)]
+to_add = [["Scotland", "SCT", "", ""]+stats,
```
You should try to find other lists to compare our list of countries too. We missed these which are not very uncommon countries so this might indicate we are missing lesser-known countries.

Christoph98 replied:
It's true that would be a better way to do it. We missed them because I figured we didn't need to add anything to the database from countries that are just going to be NaNs, thinking that the front end would simply display something like "No data available for this country" if the database didn't return them anything. We are using CIA data, so actually these countries aren't being neglected, but rather they're seen as parts of other countries -- which actually causes incoherencies. I hadn't thought about that before, but I will mention it to the team on slack. 

As far as the code that is here goes, it is only a few countries we needed to add, and due to the end of the semester I am a bit preoccupied with other things at the moment. I'll add a comment for the future group that they should probably write additional countries as csv data.

thyo9470 commented on code/routers/country_page/country.js:
```diff
@@ -99,28 +99,196 @@ router.get('/get-country-info', (req, res) => {
+/*
+    Insert Organizations API
+    inputs:
+      [{
```
The brackets here are unnecessary. Look at how the input comments are formated for the report-trips function

thyo9470 commented on code/routers/country_page/country.js:
```diff
@@ -99,28 +99,196 @@ router.get('/get-country-info', (req, res) => {
+/*
+    Remove Organizations API
+    inputs:
+      [{
```
Same as above

JoyMace commented:
Christoph none of what is in this pull request makes sense for the actual implementation of Organizations. I am closing this pull request because too much needs to be changed. I will either edit the model myself or you can contact me and we can work on it together, but this is not what we need out of this model at all.
